framework/src/utils/SymmetricRankFourTensor.C

```diff
@@ -0,0 +1,69 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "DualRealOps.h"
+#include "SymmetricRankFourTensorImplementation.h"
+
+template class SymmetricRankFourTensorTempl<Real>;
+template class SymmetricRankFourTensorTempl<DualReal>;
+
+namespace MathUtils
+{
+template <>
+void
+mooseSetToZero<SymmetricRankFourTensor>(SymmetricRankFourTensor & v)
+{
+  v.zero();
+}
+template <>
+void
+mooseSetToZero<ADSymmetricRankFourTensor>(ADSymmetricRankFourTensor & v)
+{
+  v.zero();
+}
+}
+
+#define SymmetricRankTwoTensorMultInstantiate(TemplateClass)                                       \
+  template SymmetricRankTwoTensor SymmetricRankFourTensor::operator*(                              \
+      const TemplateClass<Real> & a) const;                                                        \
+  template ADSymmetricRankTwoTensor ADSymmetricRankFourTensor::operator*(                          \
+      const TemplateClass<Real> & a) const;                                                        \
+  template ADSymmetricRankTwoTensor SymmetricRankFourTensor::operator*(                            \
+      const TemplateClass<DualReal> & a) const;                                                    \
+  template ADSymmetricRankTwoTensor ADSymmetricRankFourTensor::operator*(                          \
+      const TemplateClass<DualReal> & a) const
+
+SymmetricRankTwoTensorMultInstantiate(SymmetricRankTwoTensorTempl);
+
+template SymmetricRankFourTensor
+SymmetricRankFourTensor::operator+(const SymmetricRankFourTensor & a) const;
+template ADSymmetricRankFourTensor
+ADSymmetricRankFourTensor::operator+(const SymmetricRankFourTensor & a) const;
+template ADSymmetricRankFourTensor
+SymmetricRankFourTensor::operator+(const ADSymmetricRankFourTensor & a) const;
+template ADSymmetricRankFourTensor
+ADSymmetricRankFourTensor::operator+(const ADSymmetricRankFourTensor & a) const;
+
+template SymmetricRankFourTensor
+SymmetricRankFourTensor::operator-(const SymmetricRankFourTensor & a) const;
+template ADSymmetricRankFourTensor
+ADSymmetricRankFourTensor::operator-(const SymmetricRankFourTensor & a) const;
+template ADSymmetricRankFourTensor
+SymmetricRankFourTensor::operator-(const ADSymmetricRankFourTensor & a) const;
+template ADSymmetricRankFourTensor
+ADSymmetricRankFourTensor::operator-(const ADSymmetricRankFourTensor & a) const;
+
+template SymmetricRankFourTensor
+SymmetricRankFourTensor::operator*(const SymmetricRankFourTensor & a) const;
+template ADSymmetricRankFourTensor
+ADSymmetricRankFourTensor::operator*(const SymmetricRankFourTensor & a) const;
+template ADSymmetricRankFourTensor
+SymmetricRankFourTensor::operator*(const ADSymmetricRankFourTensor & a) const;
+template ADSymmetricRankFourTensor
+ADSymmetricRankFourTensor::operator*(const ADSymmetricRankFourTensor & a) const;
```

Comment:
Could we the macro the opname like this?
```cpp
#define SymmetricRankTwoTensorMultInstantiate(TemplateClass, opname)                                       \
  template SymmetricRankTwoTensor SymmetricRankFourTensor::operator##opname(                              \
      const TemplateClass<Real> & a) const;                                                        \
  template ADSymmetricRankTwoTensor ADSymmetricRankFourTensor::operator##opname(                          \
      const TemplateClass<Real> & a) const;                                                        \
  template ADSymmetricRankTwoTensor SymmetricRankFourTensor::operator##opname(                            \
      const TemplateClass<DualReal> & a) const;                                                    \
  template ADSymmetricRankTwoTensor ADSymmetricRankFourTensor::operator##opname(                          \
      const TemplateClass<DualReal> & a) const
```

Reply:
yeah, you are right. I did this in `ExpressionBuilder.h` way back, but this class is also getting to the point where it would pay off.

Reply:
Or even better (so that we could reuse the macro for future tensor classes):
```cpp
#define TensorMultInstantiate(TemplateClass, opname)                                       \
  template TemplateClass TemplateClass::operator##opname(                              \
      const TemplateClass<Real> & a) const;                                                        \
  template ADType<TemplateClass> ADType<TemplateClass>::operator##opname(                          \
      const TemplateClass<Real> & a) const;                                                        \
  template ADType<TemplateClass> TemplateClass::operator##opname(                            \
      const TemplateClass<DualReal> & a) const;                                                    \
  template ADType<TemplateClass> ADType<TemplateClass>::operator##opname(                          \
      const TemplateClass<DualReal> & a) const
```

Reply:
This is how it needs to look

```cpp
#define SymmetricRankFourTensorMultInstantiate(TemplateClass, opname)                              \
  template TemplateClass<Real> SymmetricRankFourTensor::operator opname(                           \
      const TemplateClass<Real> & a) const;                                                        \
  template TemplateClass<ADReal> ADSymmetricRankFourTensor::operator opname(                       \
      const TemplateClass<Real> & a) const;                                                        \
  template TemplateClass<ADReal> SymmetricRankFourTensor::operator opname(                         \
      const TemplateClass<ADReal> & a) const;                                                      \
  template TemplateClass<ADReal> ADSymmetricRankFourTensor::operator opname(                       \
      const TemplateClass<ADReal> & a) const

SymmetricRankFourTensorMultInstantiate(SymmetricRankTwoTensorTempl, *);
SymmetricRankFourTensorMultInstantiate(SymmetricRankFourTensorTempl, *);
SymmetricRankFourTensorMultInstantiate(SymmetricRankFourTensorTempl, +);
SymmetricRankFourTensorMultInstantiate(SymmetricRankFourTensorTempl, -);
```